src/nix/hash.cc

```diff
@@ -87,18 +87,35 @@ struct CmdHashBase : Command
                     return std::make_unique<HashSink>(hashAlgo);
             };
 
-            auto path2 = PosixSourceAccessor::createAtRoot(path);
+            auto makeSourcePath = [&]() -> SourcePath {
+                return PosixSourceAccessor::createAtRoot(std::filesystem::weakly_canonical(path));
+            };
+
             Hash h { HashAlgorithm::SHA256 }; // throwaway def to appease C++
             switch (mode) {
             case FileIngestionMethod::Flat:
+            {
+                // While usually we could use the some code as for NixArchive,
+                // the Flat method needs to support FIFOs, such as those
+                // produced by bash process substitution, e.g.:
```

Comment:
IIRC it was an intentional decision not to support non-regular files like `/dev/stdin` because the NAR format requires the file size to be known before we copy the contents. While it can work in O(1) space for hashing, it would require O(n) space for `addToStore()` etc. So for consistency, it may be better to not support it anywhere.

See https://github.com/NixOS/nix/pull/10019.

Reply:
We rejected `builtins.readFile` on special files, not command line arguments.

We have no real reason not to support pipes for flat hashing, and we can support it here at a minimal cost that is well worth it for users.

Note also that this fixes a regression.

Reply:
It can be used for those prefetchers to avoid having to download a file to a temporary directory.